### PR TITLE
add support for custom batches [WIP]

### PR DESCRIPTION
**Pull request by robertodessi**
## Description
Add support for custom batches.

## Related Issue (if any)
#159 

## Motivation and Context
Some games might require additional (possibly non-tensor) data to be passed to the agents. This PR supports this by using a custom Batch class that incapsulates the existing elements of a batch + a dictionary of arbitrary data.

Backwards compatibility is provided with a function that generates a Batch element from a tuple of tensors, as this is the current format for batched data in EGG.

Please note that in this implementation a label parameter is now made optional as it is not always needed e.g. in reconstruction games.

## How Has This Been Tested?
doctest pass. UTs are failing as a change in EGG games and agents (+ existing implementations in egg/zoo) should be added

TODO:

- [x] handle concatenating the aux field of a batch when batch interactions are aggregated at the end of each epoch
- [x] change signature of games and agents forward pass

### CODE REVIEW

File: egg/core/batch.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            [0.9312, 0.2153]]), receiver_input=tensor([[0.6033, 0.7328],
+            [0.1857, 0.5101]]), aux=None)
+    >>> raw_batch = (torch.rand(2, 2))
+    >>> generate_batch_from_raw(raw_batch)
+    Batch(sender_input=tensor([[0.7545, 0.2884],
+            [0.5775, 0.0358]]), labels=None, receiver_input=None, aux=None)
+    """
+    if isinstance(raw_batch, torch.Tensor):
+        return Batch(sender_input=raw_batch)
+    return Batch(*raw_batch)
```
Comment (robertodessi) [inline egg/core/batch.py:55]: test if it's an iterable otherwise raise an exception

**Review by eugene-kharitonov — COMMENTED** [review]
just a bit of discussion points!

**Comment by robertodessi** [comment]
Thanks! I integrated the comments, it indeed looks simpler and cleaner now

File: egg/core/trainers.py [hunk 2]
```diff
@@ -168,7 +169,9 @@ def eval(self):
         self.game.eval()
         with torch.no_grad():
             for batch in self.validation_data:
-                batch = move_to(batch, self.device)
+                if not isinstance(batch, Batch):
+                    batch = Batch(*batch)
+                batch.to(self.device)
```
Comment (eugene-kharitonov) [inline egg/core/trainers.py:174]: I think here you create a copy so you need to assign 
```suggestion
                batch = batch.to(self.device)
```

File: egg/core/trainers.py [hunk 3]
```diff
@@ -203,7 +206,9 @@ def train_epoch(self):
         self.optimizer.zero_grad()
 
         for batch_id, batch in enumerate(self.train_data):
-            batch = move_to(batch, self.device)
+            if not isinstance(batch, Batch):
+                batch = Batch(*batch)
+            batch.to(self.device)
```
Comment (eugene-kharitonov) [inline egg/core/trainers.py:211]: same?

**Review by eugene-kharitonov — APPROVED** [review]
i suspect two assignments are missing, otherwise lgtm!
